k9mail/src/main/java/com/fsck/k9/message/html/HtmlConverter.java

```diff
@@ -431,23 +393,35 @@ protected static String getQuoteColor(final int level) {
      * @param outputBuffer Buffer to append linked text to.
      */
     protected static void linkifyText(final String text, final StringBuffer outputBuffer) {
-        String prepared = text.replaceAll(BITCOIN_URI_PATTERN, "<a href=\"$0\">$0</a>");
-
-        Matcher m = WEB_URL_PATTERN.matcher(prepared);
-        while (m.find()) {
-            int start = m.start();
-            if (start == 0 || (start != 0 && prepared.charAt(start - 1) != '@')) {
-                if (m.group().indexOf(':') > 0) { // With no URI-schema we may get "http:/" links with the second / missing
-                    m.appendReplacement(outputBuffer, "<a href=\"$0\">$0</a>");
-                } else {
-                    m.appendReplacement(outputBuffer, "<a href=\"http://$0\">$0</a>");
-                }
-            } else {
-                m.appendReplacement(outputBuffer, "$0");
+        int currentPos = 0;
+        Matcher matcher = URI_SCHEME.matcher(text);
+
+        while (matcher.find(currentPos)) {
+            int startPos = matcher.start();
+
+            // Append all text in between
```

Comment:
We try to avoid comments because they're easily out of date when the code is changed but not the comments. It also encourages writing more readable code.

To make this more readable you could change it to

```java
String textBeforeMatch = text.substring(currentPos, startPos);
outputBuffer.append(textBeforeMatch);
```